cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ZookeeperClusterTest.java

```diff
@@ -0,0 +1,141 @@
+        assertEquals(ZookeeperCluster.LEADER_ELECTION_PORT_NAME, headless.getSpec().getPorts().get(2).getName());
+        assertEquals(new Integer(ZookeeperCluster.LEADER_ELECTION_PORT), headless.getSpec().getPorts().get(2).getPort());
+        assertEquals("TCP", headless.getSpec().getPorts().get(0).getProtocol());
+    }
+
+    @Test
+    public void testGenerateStatefulSet() {
+        // We expect a single statefulSet ...
+        StatefulSet ss = zc.generateStatefulSet(true);
+        checkStatefulSet(ss);
+    }
+
+    private void checkStatefulSet(StatefulSet ss) {
+        assertEquals(ZookeeperCluster.zookeeperClusterName(cluster), ss.getMetadata().getName());
+        // ... in the same namespace ...
+        assertEquals(namespace, ss.getMetadata().getNamespace());
+        // ... with these labels
+        assertEquals(labels("strimzi.io/cluster", cluster,
+                "strimzi.io/type", "kafka",
+                "my-user-label", "cromulent",
+                "strimzi.io/name", ZookeeperCluster.zookeeperClusterName(cluster)),
+                ss.getMetadata().getLabels());
+
+        assertEquals(new Integer(replicas), ss.getSpec().getReplicas());
+        assertEquals(image + "-zk", ss.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
+        assertEquals(new Integer(healthTimeout), ss.getSpec().getTemplate().getSpec().getContainers().get(0).getLivenessProbe().getTimeoutSeconds());
+        assertEquals(new Integer(healthDelay), ss.getSpec().getTemplate().getSpec().getContainers().get(0).getLivenessProbe().getInitialDelaySeconds());
+        assertEquals(new Integer(healthTimeout), ss.getSpec().getTemplate().getSpec().getContainers().get(0).getReadinessProbe().getTimeoutSeconds());
+        assertEquals(new Integer(healthDelay), ss.getSpec().getTemplate().getSpec().getContainers().get(0).getReadinessProbe().getInitialDelaySeconds());
+        assertEquals("foo=bar\n", AbstractModel.containerEnvVars(ss.getSpec().getTemplate().getSpec().getContainers().get(0)).get(ZookeeperCluster.ENV_VAR_ZOOKEEPER_CONFIGURATION));
+    }
+
+    /**
+     * Check that a ZookeeperCluster from a statefulset matches the one from a ConfigMap
+     */
+    @Test
+    public void testClusterFromStatefulSet() {
+        StatefulSet ss = zc.generateStatefulSet(true);
+        ZookeeperCluster zc2 = ZookeeperCluster.fromAssembly(ss, namespace, cluster);
+        // Don't check the metrics CM, since this isn't restored from the StatefulSet
+        checkService(zc2.generateService());
+        checkHeadlessService(zc2.generateHeadlessService());
+        checkStatefulSet(zc2.generateStatefulSet(true));
+    }
+
+    // TODO test volume claim templates
+
+    @Test
+    public void testPodNames() {
+
+        for (int i = 0; i < replicas; i++) {
+            assertEquals(ZookeeperCluster.zookeeperClusterName(cluster) + "-" + i, zc.getPodName(i));
```

Comment:
I think we should have `ZookeeperCluster.podName(cluster, i)`, if we haven't got it already. Likewise for the other clusters.